src/v/cluster/rm_stm.cc

```diff
@@ -86,6 +86,32 @@ static model::record_batch make_fence_batch(
     return std::move(builder).build();
 }
 
+static model::record_batch make_fence_batch_v3(
+  model::producer_identity pid,
+  model::tx_seq tx_seq,
+  std::chrono::milliseconds transaction_timeout_ms,
+  model::partition_id tm) {
+    iobuf key;
+    auto pid_id = pid.id;
+    reflection::serialize(key, model::record_batch_type::tx_fence, pid_id);
+
+    iobuf value;
+    reflection::serialize(
```

Comment:
As I understand the main obstacles are:

  - the key byte representation must not change because it's used in compaction
  - the first n bytes in value representation must not change because they are used for versioning of the format

I don't know serde well enough to recommend a nice way to overcome those obstacles. It would be nice if you could recommend something without a big increase in scope and testing area or... we may postpone this problem.

Reply:
the issue here isn't so much that we need to change to serde (e.g. `the key byte representation must not change because it's used in compaction` is a good reason for leaving it as-is). the primary issue is that (1) reflection is deprecated--we are actively removing its usage from the tree and (2) this PR is adding new reflection code but without any explanation. when someone doing reflection-code-cleanup arrives here, they are going to have to go track down all those involved for the context. all of that context should be in the commit message.

Reply:
@ZeDRoman please drop a comment e.g. `the key byte representation must not change because it's used in compaction`